File: src/main/java/com/anod/appwatcher/backup/ExportTask.java
package com.anod.appwatcher.backup;

import android.content.ContentResolver;
import android.content.Context;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;

import java.io.File;

public class ExportTask extends AsyncTask<Uri, Void, Integer> {

    private Context mContext;
    private Listener mListener;

    public ExportTask(Context context, Listener listener) {
        mContext = context;
        mListener = listener;
    }

    public interface Listener {
        void onExportStart();
        void onExportFinish(int code);
    }

    @Override
    protected void onPreExecute() {
        mListener.onExportStart();
    }

    protected Integer doInBackground(Uri... dest) {
        Uri destUri = dest[0];
        ListBackupManager mBackupManager = new ListBackupManager(mContext);
        if (destUri.getScheme().equals(ContentResolver.SCHEME_FILE))
        {
            int res = validateFileDestination(destUri);
            if (res != ListBackupManager.RESULT_OK) {
                return res;
            }
        }
        return mBackupManager.doExport(destUri);
    }

    private int validateFileDestination(Uri destUri) {
        if (!checkMediaWritable()) {
            return ListBackupManager.ERROR_STORAGE_NOT_AVAILABLE;
        }

        File destFile = new File(destUri.getPath());
        if (!destFile.getParentFile().exists()) {
            destFile.getParentFile().mkdirs();
        }

        return ListBackupManager.RESULT_OK;
    }

    /**
     * Checks if it possible to write to the backup directory
     *
     * @return true/false
     */
    private boolean checkMediaWritable() {
        String state = Environment.getExternalStorageState();
        if (!Environment.MEDIA_MOUNTED.equals(state)) {
            // We can read and write the media
            return false;
        }
        return true;
    }



    protected void onPostExecute(Integer result) {
        mListener.onExportFinish(result);
    }

}

File: app/src/main/java/com/anod/appwatcher/backup/ExportTask.kt
package com.anod.appwatcher.backup

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.os.Environment
import android.widget.Toast
import androidx.core.net.toFile
import com.anod.appwatcher.R
import info.anodsplace.framework.app.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ExportTask(private val context: ApplicationContext) {

    companion object {
        fun showFinishToast(ctx: Context, result: Int) {
            when (result) {
                DbBackupManager.RESULT_OK -> Toast.makeText(
                        ctx,
                        R.string.export_done,
                        Toast.LENGTH_SHORT
                ).show()
                DbBackupManager.ERROR_STORAGE_NOT_AVAILABLE -> Toast.makeText(
                        ctx,
                        R.string.external_storage_not_available,
                        Toast.LENGTH_SHORT
                ).show()
                DbBackupManager.ERROR_FILE_WRITE -> Toast.makeText(
                        ctx,
                        R.string.failed_to_write_file,
                        Toast.LENGTH_SHORT
                ).show()
            }
        }
    }

    suspend fun execute(destUri: Uri): Int = withContext(Dispatchers.IO) {
        if (destUri.scheme == ContentResolver.SCHEME_FILE) {
            val res = validateFileDestination(destUri)
            if (res != DbBackupManager.RESULT_OK) {
                return@withContext res
            }
        }
        return@withContext DbBackupManager(context.actual).doExport(destUri)
    }

    private fun validateFileDestination(destUri: Uri): Int {
        if (!checkMediaWritable()) {
            return DbBackupManager.ERROR_STORAGE_NOT_AVAILABLE
        }

        val destFile = destUri.toFile()
        if (destFile.parentFile?.exists() == false) {
            destFile.parentFile!!.mkdirs()
        }

        return DbBackupManager.RESULT_OK
    }

    /**
     * Checks if it possible to write to the backup directory

     * @return true/false
     */
    private fun checkMediaWritable(): Boolean {
        val state = Environment.getExternalStorageState()
        if (Environment.MEDIA_MOUNTED != state) {
            // We can read and write the media
            return false
        }
        return true
    }
}
